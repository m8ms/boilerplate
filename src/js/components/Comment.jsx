import React from 'react';
import moment from 'moment';


class Comment extends React.Component {

    getBemClasses() {
        let classes = 'comment';

        if (this.props.modifiers) {
            for (let modifier of this.props.modifiers) {
                classes += ` ${classes}--${modifier}`;
            }
        }

        return classes;
    }

    printPoints() {
        const points = this.props.comment.points;

        if (points) {
            return <div className="comment__points">{points + 'pt' + (points > 1 ? 's' : '')}</div>;
        }
    }

    printDate() {

        if (this.props.comment.datetime) {

            const date = moment(this.props.comment.datetime * 1000);

            if (date.isValid()) {
                return <div className="comment__date">{date.fromNow()}</div>;
            }
        }
    }

    printSubcomments() {
        const subcomments = this.props.comment.children || [];

        return subcomments.map(comment => <Comment modifiers={['sub-comment']} key={comment.id} comment={comment}/>);
    }

    getSubCommentsSectionClasses(){
        let classes = 'comment__sub-comments';

        if (!this.props.comment.children || !this.props.comment.children.length) {
            classes += ' comment__sub-comments--empty';
        }
        return classes;
    }

    render() {
        const comment = this.props.comment;
        return (
            <div className={this.getBemClasses()}>
                <div className="comment__author">
                    <div className="comment__username">{comment.author}</div>
                    {this.printPoints()}
                    {this.printDate()}
                </div>
                <div className="comment__content">{comment.comment}</div>
                <div className={this.getSubCommentsSectionClasses()}>
                    {this.printSubcomments()}
                </div>
            </div>
        )
    }
}

export default Comment;
