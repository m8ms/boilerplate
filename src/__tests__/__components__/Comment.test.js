import React from 'react';
import Comment from '../../js/components/Comment';
import renderer from 'react-test-renderer';
import { comment, commentWithChildren } from '../__fixtures__/comments.fixtures';

it('renders correctly with no children', () => {

    const tree = renderer
        .create(<Comment comment={comment}/>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders correctly with children', () => {

    const tree = renderer
        .create(<Comment comment={commentWithChildren}/>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
