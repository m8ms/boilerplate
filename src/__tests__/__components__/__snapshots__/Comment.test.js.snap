// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with children 1`] = `
<div
  className="comment"
>
  <div
    className="comment__author"
  >
    <div
      className="comment__username"
    >
      whalesurfertoo
    </div>
    <div
      className="comment__points"
    >
      5pts
    </div>
    <div
      className="comment__date"
    >
      14 days ago
    </div>
  </div>
  <div
    className="comment__content"
  >
    Always upvote Poland ball
  </div>
  <div
    className="comment__sub-comments"
  >
    <div
      className="comment comment--sub-comment"
    >
      <div
        className="comment__author"
      >
        <div
          className="comment__username"
        >
          ThatOneGeographyNerd
        </div>
        <div
          className="comment__points"
        >
          5pts
        </div>
        <div
          className="comment__date"
        >
          14 days ago
        </div>
      </div>
      <div
        className="comment__content"
      >
        Unless it has advanced tools. Then no. xd
      </div>
      <div
        className="comment__sub-comments"
      >
        <div
          className="comment comment--sub-comment"
        >
          <div
            className="comment__author"
          >
            <div
              className="comment__username"
            >
              whalesurfertoo
            </div>
            <div
              className="comment__points"
            >
              1pt
            </div>
            <div
              className="comment__date"
            >
              14 days ago
            </div>
          </div>
          <div
            className="comment__content"
          >
            True
          </div>
          <div
            className="comment__sub-comments comment__sub-comments--empty"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with no children 1`] = `
<div
  className="comment"
>
  <div
    className="comment__author"
  >
    <div
      className="comment__username"
    >
      SeaPrince
    </div>
    <div
      className="comment__points"
    >
      1pt
    </div>
    <div
      className="comment__date"
    >
      21 days ago
    </div>
  </div>
  <div
    className="comment__content"
  >
    I think Mexico likes Canada in this situation and called the US bumfuckers.
  </div>
  <div
    className="comment__sub-comments comment__sub-comments--empty"
  />
</div>
`;
